### PR TITLE
8312469: Print huge page information in error reporting

### PR DESCRIPTION
**Pull request by tstuefe**
We query the OS about hugepage and THP support and have a good printout function already; lets use that for hs-err and VM.info.

We can remove some of the now redundant printing. The new printout is also less wordy.

Example printout:

```
Static hugepage support:
  hugepage size: 2M
  hugepage size: 1G
  default hugepage size: 2M
Transparent hugepage (THP) support:
  THP mode: always
  THP pagesize: 2M
```

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312469](https://bugs.openjdk.org/browse/JDK-8312469): Print huge page information in error reporting (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14958/head:pull/14958` \
`$ git checkout pull/14958`

Update a local copy of the PR: \
`$ git checkout pull/14958` \
`$ git pull https://git.openjdk.org/jdk.git pull/14958/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14958`

View PR using the GUI difftool: \
`$ git pr show -t 14958`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14958.diff">https://git.openjdk.org/jdk/pull/14958.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14958#issuecomment-1645016975)

### CODE REVIEW

**Comment by tstuefe** [comment]
Friendly ping... still need a second review. Its really trivial, too.

**Comment by dholmes-ora** [comment]
Second review? ;-)

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -2141,13 +2141,6 @@ void os::Linux::print_proc_sys_info(outputStream* st) {
 void os::Linux::print_system_memory_info(outputStream* st) {
   _print_ascii_file_h("/proc/meminfo", "/proc/meminfo", st, false);
   st->cr();
-
-  // some information regarding THPs; for details see
-  // https://www.kernel.org/doc/Documentation/vm/transhuge.txt
-  _print_ascii_file_h("/sys/kernel/mm/transparent_hugepage/enabled",
-                      "/sys/kernel/mm/transparent_hugepage/enabled", st);
-  _print_ascii_file_h("/sys/kernel/mm/transparent_hugepage/defrag (defrag/compaction efforts parameter)",
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:2149]: Is it not still interesting to see these kernel values?

**Review by dholmes-ora — COMMENTED** [review]
> The new printout is also less wordy

Is it? The old printout had one line to shows all available pagesizes whilst the new has one line per available pagesize.

I'm not sure replacing the existing output is the right thing to do though ...

**Comment by tstuefe** [comment]
> > The new printout is also less wordy
> 
> Is it? The old printout had one line to shows all available pagesizes whilst the new has one line per available pagesize.
> 
> I'm not sure replacing the existing output is the right thing to do though ...

**Comment by tstuefe** [comment]
> > The new printout is also less wordy
> 
> Is it? The old printout had one line to shows all available pagesizes whilst the new has one line per available pagesize.
> 
> I'm not sure replacing the existing output is the right thing to do, though ...

Thanks David.

Sorry, my fault; my description has been imprecise.

The old printout printed:

1) two THP parameters, of which the second seemed a bit arbitrary (there would be more interesting candidates, e.g. `/shmem_enabled`), verbatim:

```
/sys/kernel/mm/transparent_hugepage/enabled: always [madvise] never                                                                                                                                                                                                             
/sys/kernel/mm/transparent_hugepage/defrag (defrag/compaction efforts parameter): always defer defer+madvise [madvise] never
```

2) the *curated* list of large pages, if the VM is called with large pages, as part of the "OS Memory" section

```
Memory: 4k page, physical 65773476k(59463004k free), swap 2097148k(2097148k free)
Page Sizes: 4k
```

This is not optimal because the VM has two lists of page sizes: the "real" one - reflecting the list of page sizes at the OS level - and a curated one that only contains page sizes the JVM is allowed to use. At (2) we only print the latter. For example, if we start the JVM without `+UseLargePages`, the printed list will only show "4K". So, the printed page list is not "the truth" and cannot be used to analyze erroneous OS configuration setups.

`/sys/kernel/mm/transparent_hugepage/enabled: always [madvise] never` is, in the new printout, identical to the `THP mode` printout.

But I see what you mean with wordy. I'll withdraw the PR and massage it a bit to condense the information.

**Comment by tstuefe** [comment]
Exhumed to finally finish this issue.

@dholmes-ora: back in August you argued that the output was too wordy for the hs-err file. I reworked the printing routine to produce more condensed output. New output looks like this:

```
Explicit hugepage support: [2M],1G
THP support:               madvise-enabled, 2M
Shared memory THP support: never
```

Unfortunately, this made the patch a lot more complex, mainly because the parser in the jtreg test had to be rewritten. 

I also removed information from the raw-printing-of-config-file-section that are now redundant. So the resulting hs-err file size is actually smaller now.
